File: py_drill/Problems/PQ03_KthLargestElementInStream.py
# REF : https://leetcode.com/problems/kth-largest-element-in-a-stream/
# NOTES :
#       Kth Largest element using minHeap

# T: O(logn) of add, O( (n-k)*logn ) of constructor ~ O(nlogn)
# S: O(n) due to minHeap

import heapq

class KthLargest:

    def __init__(self, k: int, nums: list[int]):
        self.k = k
        self.minHeap = nums
        heapq.heapify(self.minHeap)
        while len(self.minHeap) > k:
            heapq.heappop(self.minHeap)

    def add(self, val: int) -> int:
        heapq.heappush(self.minHeap, val)
        if len(self.minHeap) > self.k:
            heapq.heappop(self.minHeap)
        return self.minHeap[0]


# Your KthLargest object will be instantiated and called as such:
# obj = KthLargest(k, nums)
# param_1 = obj.add(val)
def main():
    print("output : {}".format(KthLargest(3, [4, 5, 8, 2]).add(3)))

if __name__ == "__main__":
    main()

File: cpp_drill/Problems/PQ03_KthLargestElementInStream.cpp
#include <iostream>
#include <queue>
using namespace std;

// REF : https://leetcode.com/problems/kth-largest-element-in-a-stream/
// NOTES :
//      Same logice like PQ02_KthLargestElementInAnArray
//      Push nums element into Min Heap of len k. Pop elements exceeding k.
//      In Add method, add element into Min Heap and return top element AND
//      if k is exceeding then pop the element from Min Heap.

// T: O(n*logk)
// S: O(k)

class KthLargest {
private:
    int k;
    priority_queue<int, vector<int>, greater<int>> minHeap;
public:
    KthLargest(int k, vector<int>& nums) {
        this->k = k;
        for (int &i : nums) {
            minHeap.push(i);

            if (minHeap.size() > k)
                minHeap.pop();
        }
    }
    
    int add(int val) {
        minHeap.push(val);
        if (minHeap.size() > k)
            minHeap.pop();
        
        return minHeap.top();
    }
};

/**
 * Your KthLargest object will be instantiated and called as such:
 * KthLargest* obj = new KthLargest(k, nums);
 * int param_1 = obj->add(val);
 */